lib/source_file.h

```diff
@@ -168,7 +168,9 @@ class SourceInfo final {
         return *this;
     }
 
-    bool operator==(const SourceInfo &rhs) const { return start == rhs.start && end == rhs.end; }
+    bool operator==(const SourceInfo &rhs) const {
+        return start == rhs.start && end == rhs.end && getSourceFile() == rhs.getSourceFile();
```

Comment:
The source file depends (solely) on the (start) line number -- so start == rhs.start, then the source files by definition must be identical.  So this extra comparison can never add anything?

Reply:
Hey, @ChrisDodd, can you please take a look at https://github.com/p4lang/p4c/issues/4026, maybe you have a suggestion.

Reply:
> The source file depends (solely) on the (start) line number -- so if start == rhs.start, then the source files by definition must be identical. So this extra comparison can never add anything?

How is sourceInfo constructed? Or rather how is the file name stored?

I have been running into issues with #include files and their source info. 

I am maintaining a map which compares the sourceInfo of IR nodes here: 
https://github.com/p4lang/p4c/pull/4029/files#diff-0c31e90ea5ad25031d5734ed00bc8dee477c6c54a3208d1476b26d1eb5407c33R10
This works reasonably well, but in some cases the sourceLine line and column number are the same, but the originating file is different, which seems to lead to clashes.